### ISSUE TITLE
Build number exceeds version format

### ISSUE THREAD

**Issue by Krzmbrzl** [mumble-voip/mumble#5827]
### Description

In the latest stable release (1.4.274) the client will advertise its version as 1.5.18 to the server. This causes issues in case one is connecting to a server that is running Mumble 1.5 or more recent as in that version a new audio protocol has been introduced that is used for all clients advertising themselves as 1.5.0 or higher. However, as the 1.4.274 doesn't know about this new protocol, this will effectively cause a loss of audio for that client.

### Steps to reproduce

1. Connect to any server
2. Check the reported client version on that server

### Mumble version

1.4.274

### Mumble component

Both

### OS

Linux

### Reproducible?

Yes

### Additional information

The issue boils down to how we encode version numbers as integers:
https://github.com/mumble-voip/mumble/blob/789f2d79e1b83db88e053c6ec8af93d885350a9c/src/Version.cpp#L42

From this it can be seen that we divide an integer (in the most general case a 32-bit integer) into equal parts: 8 bits encoding the major version number, 8 bits for the minor version number and 8 bits for the patch number (which we substitute with the build number).

That means that each of these fields can have values in the range 0-255.

In 1.4.274 the build (patch) number exceeds this range and therefore produces an overflow in this scheme, incrementing the minor version number by one, leading to the observed result.

**Comment by Krzmbrzl** [comment]
In order to fix this, I see only 2 choices:
1. Change the versioning scheme (again) to ensure that we never encounter version numbers >255
2. Change the version encoding scheme to allow for larger patch numbers (at the cost of the max value of the other version numbers)

Since I wouldn't want to change the versioning scheme again, I think the only viable option is 2. However, this has to somehow be done in a backwards compatible way in order to not screw up the communication with older clients/servers.

**Comment by Hartmnt** [comment]
I have not yet looked at the code, yet. If we are lucky, the server advertises its version before the client in the handshake. That would mean a >= 1.4.(274+1) client could conditionally use a new version scheme for a >= 1.4.(274+1) server and report '1.4.255' for all previous server versions, which should be backward compatible as long as patches do not contain breaking changes.

If the client advertises before the server... well that would probably require a hacky workaround.

**Comment by ZeroAbility** [comment]
What happened to this:

https://docs.microsoft.com/en-us/windows/win32/msi/version

Does Linux and macOS not respect this version scheme? The reason for the change initially was because of that and upgrade issues with the installer, I thought. We dropped the final segment because it breaks upgrades for MSI's. I guess I don't understand the reason for the 8 bit type instead of the 16 bit one.

**Comment by davidebeatrici** [comment]
Exactly, that is the problem: a limitation of our protocol.

If we're lucky we can just shift 1 byte and use it for the third component (build number).

**Comment by Hartmnt** [comment]
I have looked at the current situation:

The client sends its version to the server [here](https://github.com/mumble-voip/mumble/blob/master/src/mumble/ServerHandler.cpp#L787)

The server sends its version to the client [here](https://github.com/mumble-voip/mumble/blob/master/src/murmur/Server.cpp#L1545)

It appears these messages are sent independently, as soon as the connection is established in an non-deterministic order.

Furthermore, every use of ``uiVersion`` is also affected by this. [set here](https://github.com/mumble-voip/mumble/blob/master/src/mumble/ServerHandler.cpp#L200) and [used for example here](https://github.com/mumble-voip/mumble/blob/master/src/mumble/MainWindow.cpp#L1705)
These would probably have to be updated by using something like [this](https://github.com/mumble-voip/mumble/blob/master/src/mumble/ACLEditor.cpp#L152)

**Comment by davidebeatrici** [comment]
A potential solution I proposed in our team chat would be to treat the "blob" sent as the protocol version.

There is no technical requirement in knowing the specific software (client/server) version as I'm pretty sure all checks we have in the code strictly refer to the protocol one.

Also, this change would be beneficial for the eventual switch to libmumble.

**Comment by Hartmnt** [comment]
Another thing: The version message parsing in [client](https://github.com/mumble-voip/mumble/blob/master/src/mumble/Messages.cpp#1158) and [server](https://github.com/mumble-voip/mumble/blob/master/src/murmur/Messages.cpp#L1924) seem to be, at first glance, without side effect.

That means the that new server and clients could initially send 1.4.255 and after learning that the other side also reports 1.4.255 simply resend the version message again with an updated version format.

**Comment by davidebeatrici** [comment]
Excellent idea.

**Comment by Hartmnt** [comment]
Quick implementation question: Is the order of optional fields in ``src/Mumble.proto`` important? I assume it is not.

**Comment by Krzmbrzl** [comment]
The order itself is not important, but the assigned ID (the numeric value assigned to the field) is important and must not be changed. And it kinda makes sense to have the fields ordered based on their ID :shrug:

**Comment by Hartmnt** [comment]
> The order itself is not important, but the assigned ID (the numeric value assigned to the field) is important and must not be changed. And it kinda makes sense to have the fields ordered based on their ID shrug

Yeah, that's what I figured. At first I was thinking to add a separate message for patch information, but since the fields are all optional, we might as well add an optional field to the regular version packet. And aesthetically it would have been nice to put the patch right after the version, but that would break compatibility i think... Anyway I added it to the bottom

**Comment by Hartmnt** [comment]
Before I have to search for a long time: Where do I set the patch version in the build pipeline for testing purposes? (I can only build the 1.5 branch, so I want to set the patch level manually to >255)

**Comment by makedir** [comment]
My mumble server is still

![image](https://github.com/mumble-voip/mumble/assets/6721083/f7fb7369-4db2-48a2-b825-85b943a208fa)

can someone here maybe help me where or how I can get a newer package of mumble-server for Ubuntu 20.04.6?

**Comment by Hartmnt** [comment]
@makedir This is very much the wrong issue for this topic, but here is a rundown:

The PPA is unmaintained, but it will be again eventually. There is no ETA for that though. We also did not manage to get 1.5 into the Debian package freeze, which means it will take a long time to drip down to Debian based distros. For the foreseeable future, you will have to compile Mumble yourself, use a docker image, or stick with the old version.

PPA: https://github.com/mumble-voip/mumble-ubuntu-ppa/issues/16
Docker: https://github.com/mumble-voip/mumble-docker
